### ISSUE TITLE
Properties on Y.Map can't be accessed with .get before map is attached to doc

### ISSUE THREAD

**Issue by SamyPesse** [yjs/yjs#666]
When defining properties on a `Y.Map` that has not been attached to a `Y.Doc` yet, the properties can't be accessed using `.get`.

**Steps to reproduce**

Code to reproduce:

```js
const map = new Y.Map();
map.set('prop', 'value')

const result = map.get('prop');
// Expect `result` to be `value`, but instead gets `undefined`
```

**Expected behavior**

It's unclear if it's a bug or by design. But I felt like an issue should be opened as if it's by design, this should be documented in the API docs.

On `Y.Array`, it looks like the behaviour is even harder to anticipate:
- [`.length` takes `prelimContent` in consideration](https://github.com/yjs/yjs/blob/main/src/types/YArray.js#L107) and will return the right length before the array is attached to the document.
- `.get`, etc don't take `prelimContent` in consideration

I think, one of the following should be done:

- Clarify in the API documentation of each methods how they behave before/after being attached
- Fix the different methods to return consistently before/after being attached to a document

**Additional context**

* [x] I'm a sponsor 💖
* [ ] This issue is a blocker for my project.

**Comment by dmonad** [comment]
Good catch!

The methods are not supposed to return any content before the type is integrated.

I updated the documentation https://docs.yjs.dev/getting-started/working-with-shared-types#caveats to mention that types will appear empty until they are integrated.

Also, I updated the `.length` property to return 0 if the type is not yet integrated.

I really should rework the documentation.. It doesn't to justice to the awesome tool that I'm using :)

**Comment by SamyPesse** [comment]
Thanks @dmonad for the update!

I'm curious, what is the reason for this limit?

And I might suggest 2 changes to improve the DX around yjs:

1. Throw an error in `.get()`, `.length`, etc; instead of returning `null/undefined/0`. Returning a value seems to indicate that it can be used, while it should not. An error could help directly identify during development invalid cases.
2. For the JSON copy problem, similar to `recoil`, I think we should do a `if (process.env.NODE_ENV === 'development') { deepFreeze(value) }`

These 2 things could help prevent errors that are very hard to catch and identify; because not everyone in a team using yjs will have a perfect understanding of the caveats.

At the moment, for GitBook, we are tracking some bugs about loss of contents that might be related to this problem, but since it still returns something it's very hard to identify where the issue could be in our code. We'll probably patch yjs to throw an error until we get to the bottom of it.

**Comment by dmonad** [comment]
The data is represented using the CRDT data structure. Without the connected Yjs document we can't generate unique ids etc.. 

I could maintain two different representations (i.e. add a representation of the data that doesn't require the Yjs document). But then I would also need to add support for events, etc.. 

Hence I opted to make the data types write-only until you add the data to a Yjs document.

Your suggestions make sense. I added a ticket #667 to track progress. I hope that I can get to it next week.
